src/java/org/apache/cassandra/tcm/AbstractLocalProcessor.java

```diff
@@ -53,19 +54,22 @@ public final Commit.Result commit(Entry.Id entryId, Transformation transform, fi
     {
         while (!retryPolicy.reachedMax())
         {
-            ClusterMetadata previous = log.waitForHighestConsecutive();
-            if (!previous.fullCMSMembers().contains(FBUtilities.getBroadcastAddressAndPort()))
+            long timestampMicros = -1;
+            log.waitForHighestConsecutive();
+            LocalLog.CommittedValue prevCommitted = log.committedValue();
+            if (!prevCommitted.metadata.fullCMSMembers().contains(FBUtilities.getBroadcastAddressAndPort()))
                 throw new IllegalStateException("Node is not a member of CMS anymore");
             Transformation.Result result;
             if (!CassandraRelevantProperties.TCM_ALLOW_TRANSFORMATIONS_DURING_UPGRADES.getBoolean() &&
                 !transform.allowDuringUpgrades() &&
-                previous.metadataSerializationUpgradeInProgress())
+                prevCommitted.metadata.metadataSerializationUpgradeInProgress())
             {
                 result = new Transformation.Rejected(ExceptionCode.INVALID, "Upgrade in progress, can't commit " + transform);
             }
             else
             {
-                result = transform.execute(previous);
+                while ((timestampMicros = nextUnixMicros()) <= prevCommitted.timestampMicros);
```

Comment:
Could we do something better than active wait? like sleep?
perhaps it makes sense to `WARN` if the difference is too big?

Reply:
Indeed, we can either wait or assume the previous epoch timestamp + 1 in case the current timestamp is lower than the last epoch timestamp. In both cases we can warn that clocks of CMS nodes have diverged too much.